core/unit_test/TestAtomics.hpp

```diff
@@ -539,4 +539,45 @@ TEST(TEST_CATEGORY, atomics) {
 #endif
 }
 
+// see https://github.com/trilinos/Trilinos/pull/11506
+struct TpetraUseCase {
+  template <class Scalar>
+  struct WrapScalarAndCompareAbsMax {
+    Scalar value;
+
+   private:
+    friend KOKKOS_FUNCTION bool operator<(
+        WrapScalarAndCompareAbsMax const& lhs,
+        WrapScalarAndCompareAbsMax const& rhs) {
+      return Kokkos::abs(lhs.value) < Kokkos::abs(rhs.value);
+    }
+    friend KOKKOS_FUNCTION bool operator>(
+        WrapScalarAndCompareAbsMax const& lhs,
+        WrapScalarAndCompareAbsMax const& rhs) {
+      return Kokkos::abs(lhs.value) > Kokkos::abs(rhs.value);
+    }
+  };
+
+  using T = int;
+  Kokkos::View<T> d_{"lbl"};
+  KOKKOS_FUNCTION void operator()(int i) const {
+    // 0, -1, 2, -3, ...
+    auto v_i = static_cast<T>(i);
+    if (i % 2 == 1) v_i = -v_i;
+    Kokkos::atomic_max(reinterpret_cast<WrapScalarAndCompareAbsMax<T>*>(&d_()),
+                       WrapScalarAndCompareAbsMax<T>{v_i});
```

Comment:
```suggestion
    WrapScalarAndCompareAbsMax<T> tmp{d_()};
    Kokkos::atomic_max(&tmp, WrapScalarAndCompareAbsMax<T>{v_i});
    d_() = tmp.value;
```
to avoid undefined behavior, also see https://stackoverflow.com/questions/62765714/is-it-safe-to-reinterpret-cast-between-a-type-and-a-wrappertype. The `reinterpret_cast` is only guaranteed to have the expected behavior if `d_()` would already be wrapped by the temporary `WrapScalarAndCompareAbsMax`.

Reply:
Would not be atomic then.

Reply:
Hmmm, yes, that's right.

Reply:
Doesn't that mean that the functor should use `WrapScalarAndCompareAbsMax` directly then? Or do we think it's reasonable for the compiler to support this technically undefined behavior in the expected way?

Reply:
> Doesn't that mean that the functor should use `WrapScalarAndCompareAbsMax` directly then?

Sure but that would not help for the the Tpetra use case.

> Or do we think it's reasonable for the compiler to support this technically undefined behavior in the expected way?

Acknowledging that it is bad but expecting that it will work

Reply:
```c++
 Kokkos::atomic_max(reinterpret_cast<WrapScalarAndCompareAbsMax<T>*>(&d_()),
                       WrapScalarAndCompareAbsMax<T>{v_i});
```

I ... didn't write that.  Just gonna put that out there in case somebody claims it's my fault ; - P